google/cloud/spanner_dbapi/_helpers.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2020 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+from google.cloud.spanner_dbapi.parse_utils import get_param_types
+from google.cloud.spanner_dbapi.parse_utils import parse_insert
+from google.cloud.spanner_dbapi.parse_utils import sql_pyformat_args_to_spanner
+from google.cloud.spanner_v1 import param_types
+
+
+SQL_LIST_TABLES = """
+            SELECT
+              t.table_name
+            FROM
+              information_schema.tables AS t
+            WHERE
+              t.table_catalog = '' and t.table_schema = ''
+            """
+
+SQL_GET_TABLE_COLUMN_SCHEMA = """SELECT
+                COLUMN_NAME, IS_NULLABLE, SPANNER_TYPE
+            FROM
+                INFORMATION_SCHEMA.COLUMNS
+            WHERE
+                TABLE_SCHEMA = ''
+            AND
+                TABLE_NAME = @table_name
+            """
+
+# This table maps spanner_types to Spanner's data type sizes as per
+#   https://cloud.google.com/spanner/docs/data-types#allowable-types
+# It is used to map `display_size` to a known type for Cursor.description
+# after a row fetch.
+# Since ResultMetadata
+#   https://cloud.google.com/spanner/docs/reference/rest/v1/ResultSetMetadata
+# does not send back the actual size, we have to lookup the respective size.
+# Some fields' sizes are dependent upon the dynamic data hence aren't sent back
+# by Cloud Spanner.
+code_to_display_size = {
+    param_types.BOOL.code: 1,
+    param_types.DATE.code: 4,
+    param_types.FLOAT64.code: 8,
+    param_types.INT64.code: 8,
+    param_types.TIMESTAMP.code: 12,
+}
+
+
+def _execute_insert_heterogenous(transaction, sql_params_list):
+    for sql, params in sql_params_list:
+        sql, params = sql_pyformat_args_to_spanner(sql, params)
+        param_types = get_param_types(params)
+        res = transaction.execute_sql(
+            sql, params=params, param_types=param_types
+        )
+        # TODO: File a bug with Cloud Spanner and the Python client maintainers
+        # about a lost commit when res isn't read from.
```

Comment:
It would be a good question for the original code developer who unfortunately is no longer a member of our team. I'm not even sure what's the purpose of `res`, as the function does not return anything. We should probably revise this method later.